### DISCUSSION TITLE
About recover data from files saved by minio

### DISCUSSION THREAD

**Discussion by p363796822** [milvus-io/milvus#18524] [Q&A and General discussion]
I have read the article on the WeChat public account: Milvus 2.0 Data Insertion and Persistence, but the introduction of binlog is too simple. The compressed_page_size parameter is missing from the parquet file, so I would like to ask how to get the data inserted into the milvus through the file in the minio.

The milvus version I am using is the latest 2.1.0

**Comment by yhmo** [comment]
The cpp code to read/write parquet file is here: internal/core/src/storage/parquet_c.cpp
At the line no.331 we can see:
```
ast = parquet::arrow::WriteTable(
            *table, mem_pool, p->output, 1024 * 1024 * 1024,
            parquet::WriterProperties::Builder().compression(arrow::Compression::ZSTD)->compression_level(3)->build());
```
I am not sure, the "compressed_page_size" may be a parameter of the parquet::WriterProperties::Builder(), right?
The constructor of parquet::WriterProperties::Builder():
```
Builder()
        : pool_(::arrow::default_memory_pool()),
          dictionary_pagesize_limit_(DEFAULT_DICTIONARY_PAGE_SIZE_LIMIT),
          write_batch_size_(DEFAULT_WRITE_BATCH_SIZE),
          max_row_group_length_(DEFAULT_MAX_ROW_GROUP_LENGTH),
          pagesize_(kDefaultDataPageSize),
          version_(ParquetVersion::PARQUET_1_0),
          data_page_version_(ParquetDataPageVersion::V1),
          created_by_(DEFAULT_CREATED_BY) {}
```
Here we can see "pagesize_(kDefaultDataPageSize)", the kDefaultDataPageSize value is 1024 * 1024.

**Reply by p363796822** [reply]
I use the python version of the fastparquet package to read the file, and the parameter I mentioned above that lacks compressed_page_size will appear. I have not tried the cpp method here.
Another question is, how do I deserialize the data that has been stored in the minio and get the written data.Can you give an example about deserialization? Thank you

**Reply by yhmo** [reply]
In fact, It is also difficult for us to deserialize milvus binlog data.  Recently we have a new feature "backup milvus data" could be related with this work, but the feature is still in design stage, not yet ready to begin implementation.

**Reply by p363796822** [reply]
Yes, I know it's really difficult, but what I want to know is whether reading parquet files from minio can achieve the purpose of getting inserted data, my current state is that I don't know how to read the files, one by one Reading or reading multiple files together, can you help me understand?

**Reply by yhmo** [reply]
As we know, for each collection, if you take it as column-based, each collection has multiple fields; if you take it as row-based, each collection has multiple segments. We can draw a matrix like this:
```
______________|_  field_1 ____|______ field_2 ____|
              |    binlog_11  |     binlog_12     |
segment_1     |    binlog_21  |     binlog_22     |
______________|_______________|___________________|
              |    binlog_31  |     binlog_32     |
segment_2     |    binlog_41  |     binlog_42     |
______________|_______________|___________________|

```
We can see each field has multiple binlog files, each segment also has multiple binlog files. The inserted data is split/persisted into binlog files. Let's say 10000 rows were inserted, the binlog_11 store rows from 1 ~ 2500 for the field_1, the binlog_12 store rows from 1 ~ 2500 for field_2, and so on, the binlog_21 store rows from 2501 ~ 5000 for the field_1, the binlog_22 store rows from 2501 ~ 5000 for field_2, etc.

In milvus, each collection/field/partition/segment/binlog has an ID, if you insert some data and browse the minio you will see the data files are managed in this structure:


```
── files
    ├── insert_log
    │   └── 434574377495035905 (collection ID)
    │       └── 434574377495035906 (partition ID)
    │           ├── 434574378976149505 (segment ID)
    │           │   ├── 0 (field ID)
    │           │   │   └── 434574379605819396 (binlog ID)
    │           │   ├── 1
    │           │   │   └── 434574379605819397
    │           │   ├── 100
    │           │   │   └── 434574379605819393
    │           │   ├── 101
    │           │   │   └── 434574379605819394
    │           │   └── 102
    │           │       └── 434574379605819395
    │           └── 434574378976149506
    │               ├── 0
    │               │   └── 434574379605819401
    │               ├── 1
    │               │   └── 434574379605819402
    │               ├── 100
    │               │   └── 434574379605819398
    │               ├── 101
    │               │   └── 434574379605819399
    │               └── 102
    │                   └── 434574379605819400

```

The binlog id is generated by timestamp. Sometimes you will see multiple binlog files under a field, then the big id is newly created.
Each binlog file is generated by the PayloadWriter class, which I mentioned in comment, a wrapper of parquet lib.
We can see this example has 5 fields, the ID is: 0, 1, 100, 101, 102. The 0 and 1 is internal fields, you can ignore them. The 100, 101, 102 are visible to client, the collection schema has 3 fields defined by user. User-defined field ID always from 100. Fox example, this collection has 3 fields, the first one is id field, the second is a bool type field, the third is a float vector field, then you read the binlog files from 100/101/102, parse them as int64/bool/floatvector respectively

**Reply by p363796822** [reply]
Thank you for your patient reply, then I understand that from the above file organization, each individual file should be readable, right? It does not need multiple files to be combined to be read. A single file should be It is readable, but the amount of data is incomplete. Do I understand much?

**Reply by yhmo** [reply]
Yes, read id list 1 ~ 2500 from binlog_11, 2501 ~ 5000 from binlog_21, 5001 ~ 7500 from binlog_31, 7501 ~ 10000 from binlog_41, then you get all data of the field_1.

**Reply by p363796822** [reply]
However, I have tried various ways to read a single parquet file. Is the way on my side wrong or what is going on? Can you read it successfully?

**Reply by yhmo** [reply]
> However, I have tried various ways to read a single parquet file. Is the way on my side wrong or what is going on? Can you read it successfully?

I didn't try read it before. Not sure what is the problem. I think you need to read the source code to know how it generate the parquet data.

**Reply by p363796822** [reply]
If the minio is not readable then can I read the previously inserted data through pulsar？

**Reply by p363796822** [reply]
I use pyarrow to read the data in minio and still can't read it, the error is：OSError: Couldn't deserialize thrift: TProtocolException: Invalid data Deserializing page header failed.
I looked at the parquet_c.cpp file in the internal, it does not exist in the 2.1.0 version, so how to write the parquet file in the 2.1.0 version?

**Reply by p363796822** [reply]
If it is convenient, please read the data file written to the minio and see if there is also a problem? Thank you very much

**Reply by yhmo** [reply]
> If the minio is not readable then can I read the previously inserted data through pulsar？

Read data from pulsar is not a good idea. Pulsar stores all insert requests as logs, datanode and querynode subscribe the data from pulsar, it is difficult to know how much data was consumed so you don't know fetch data from pulsar from which position.

**Reply by yhmo** [reply]
> I use pyarrow to read the data in minio and still can't read it, the error is：OSError: Couldn't deserialize thrift: TProtocolException: Invalid data Deserializing page header failed. I looked at the parquet_c.cpp file in the internal, it does not exist in the 2.1.0 version, so how to write the parquet file in the 2.1.0 version?

In latest source code, the parquet_c.cpp is here: https://github.com/milvus-io/milvus/blob/master/internal/core/src/storage/parquet_c.cpp

In v2.1 source code, the parquet wrapper is here: https://github.com/milvus-io/milvus/blob/2.1.0/internal/storage/cwrapper/ParquetWrapper.cpp

**Reply by p363796822** [reply]
Can I send you a file here, and you can parse it for me to see if it can be parsed?
[435007196188377089.zip](https://github.com/milvus-io/milvus/files/9265841/435007196188377089.zip)

**Reply by p363796822** [reply]
Hello, can you help me? I have tried many ways to read the parquet file, but the displayed file itself is not legal. Can you help me to find out the reason? You can also try to read the file I uploaded above. Look, it's very small and won't take up too many resources. Please help to troubleshoot the problem.Thank you.

**Reply by yhmo** [reply]
It also difficult for us to parse it... I will try it when I get time.

**Reply by yhmo** [reply]
Here is an example to parse a binlog file, written by Go: https://github.com/milvus-io/milvus/blob/master/cmd/tools/binlog/main.go
Hope it can help. You'd better read into the code to know how it works.

**Reply by p363796822** [reply]
Thank you so much, I'll try it out here
